// @pact-foundation/pact/src/v3/graphql/graphQL.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLPactV3 = void 0;
var lodash_1 = require("lodash");
var ramda_1 = require("ramda");
var configurationError_1 = require("../../common/graphQL/configurationError");
var pact_1 = require("../pact");
var types_1 = require("../../common/graphQL/types");
var matchers_1 = require("../matchers");
var graphQL_1 = require("../../common/graphQL/graphQL");
/**
 * Expose a V3 compatible GraphQL interface
 *
 * Code borrowed/inspired from https://gist.github.com/wabrit/2d1e1f9520aa133908f0a3716338e5ff
 */
var GraphQLPactV3 = /** @class */ (function (_super) {
    __extends(GraphQLPactV3, _super);
    function GraphQLPactV3() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.operation = undefined;
        _this.variables = undefined;
        _this.req = undefined;
        return _this;
    }
    GraphQLPactV3.prototype.given = function (providerState, parameters) {
        _super.prototype.given.call(this, providerState, parameters);
        return this;
    };
    GraphQLPactV3.prototype.uponReceiving = function (description) {
        _super.prototype.uponReceiving.call(this, description);
        return this;
    };
    /**
     * The GraphQL operation name, if used.
     * @param operation {string} the name of the operation
     * @return this object
     */
    GraphQLPactV3.prototype.withOperation = function (operation) {
        this.operation = operation;
        return this;
    };
    /**
     * Add variables used in the Query.
     * @param variables {GraphQLVariables}
     * @return this object
     */
    GraphQLPactV3.prototype.withVariables = function (variables) {
        this.variables = variables;
        return this;
    };
    /**
     * The actual GraphQL query as a string.
     *
     * NOTE: spaces are not important, Pact will auto-generate a space-insensitive matcher
     *
     *  e.g. the value for the "query" field in the GraphQL HTTP payload:
     *  '{ "query": "{
     *        Category(id:7) {
     *          id,
     *          name,
     *          subcategories {
     *            id,
     *            name
     *          }
     *        }
     *     }"
     *  }'
     * @param query {string|ASTNode} parsed or unparsed query
     * @return this object
     */
    GraphQLPactV3.prototype.withQuery = function (query) {
        this.query = (0, graphQL_1.validateQuery)(query, types_1.OperationType.Query);
        return this;
    };
    /**
     * The actual GraphQL mutation as a string or parse tree.
     *
     * NOTE: spaces are not important, Pact will auto-generate a space-insensitive matcher
     *
     * e.g. the value for the "query" field in the GraphQL HTTP payload:
     *
     * mutation CreateReviewForEpisode($ep: Episode!, $review: ReviewInput!) {
     *   createReview(episode: $ep, review: $review) {
     *     stars
     *     commentary
     *   }
     * }
     * @param mutation {string|ASTNode} parsed or unparsed mutation
     * @return this object
     */
    GraphQLPactV3.prototype.withMutation = function (mutation) {
        this.query = (0, graphQL_1.validateQuery)(mutation, types_1.OperationType.Mutation);
        return this;
    };
    /**
     * Used to pass in the method, path and content-type; the body detail would
     * not typically be passed here as that will be internally constructed from
     * withQuery/withMutation/withVariables calls.
     *
     * @see {@link withQuery}
     * @see {@link withMutation}
     * @see {@link withVariables}
     * @param req {V3Request} request
     * @return this object
     */
    GraphQLPactV3.prototype.withRequest = function (req) {
        // Just take what we need from the request, as most of the detail will
        // come from withQuery/withMutation/withVariables
        this.req = req;
        return this;
    };
    /**
     * Overridden as this is the "trigger point" by which we should have received all
     * request information.
     * @param res {V3Response} the expected response
     * @returns this object
     */
    GraphQLPactV3.prototype.willRespondWith = function (res) {
        var _a, _b;
        if (!this.query) {
            throw new configurationError_1.ConfigurationError('You must provide a GraphQL query.');
        }
        if (!this.req) {
            throw new configurationError_1.ConfigurationError('You must provide a GraphQL request.');
        }
        this.req = __assign(__assign({}, this.req), { body: (0, ramda_1.reject)(lodash_1.isUndefined, {
                operationName: this.operation,
                query: (0, matchers_1.regex)((0, graphQL_1.escapeGraphQlQuery)(this.query), this.query),
                variables: this.variables,
            }), headers: {
                'Content-Type': ((_a = this.req).contentType || (_a.contentType = 'application/json')),
            }, method: ((_b = this.req).method || (_b.method = 'POST')) });
        _super.prototype.withRequest.call(this, this.req);
        _super.prototype.willRespondWith.call(this, res);
        return this;
    };
    GraphQLPactV3.prototype.addInteraction = function () {
        throw new configurationError_1.ConfigurationError('Only GraphQL Queries are allowed');
    };
    GraphQLPactV3.prototype.withRequestBinaryFile = function () {
        throw new configurationError_1.ConfigurationError('Only GraphQL Queries are allowed');
    };
    GraphQLPactV3.prototype.withRequestMultipartFileUpload = function () {
        throw new configurationError_1.ConfigurationError('Only GraphQL Queries are allowed');
    };
    return GraphQLPactV3;
}(pact_1.PactV3));
exports.GraphQLPactV3 = GraphQLPactV3;
