// @pact-foundation/pact/src/v4/http/interactionWithPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionWithPlugin = void 0;
var requestWithPluginBuilder_1 = require("./requestWithPluginBuilder");
var interactionWithPluginRequest_1 = require("./interactionWithPluginRequest");
var matchers_1 = require("../../v3/matchers");
var InteractionWithPlugin = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function InteractionWithPlugin(pact, interaction, opts, cleanupFn) {
        this.pact = pact;
        this.interaction = interaction;
        this.opts = opts;
        this.cleanupFn = cleanupFn;
    }
    // Multiple plugins are allowed
    InteractionWithPlugin.prototype.usingPlugin = function (config) {
        this.pact.addPlugin(config.plugin, config.version);
        return this;
    };
    InteractionWithPlugin.prototype.withRequest = function (method, path, builder) {
        this.interaction.withRequest(method, (0, matchers_1.matcherValueOrString)(path));
        if (typeof builder === 'function') {
            builder(new requestWithPluginBuilder_1.RequestWithPluginBuilder(this.interaction));
        }
        return new interactionWithPluginRequest_1.InteractionWithPluginRequest(this.pact, this.interaction, this.opts, this.cleanupFn);
    };
    return InteractionWithPlugin;
}());
exports.InteractionWithPlugin = InteractionWithPlugin;
