// @pact-foundation/pact/src/v4/http/interactionWithPluginRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionWithPluginRequest = void 0;
var interactionWithPluginResponse_1 = require("./interactionWithPluginResponse");
var responseWithPluginBuilder_1 = require("./responseWithPluginBuilder");
var InteractionWithPluginRequest = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function InteractionWithPluginRequest(pact, interaction, opts, cleanupFn) {
        this.pact = pact;
        this.interaction = interaction;
        this.opts = opts;
        this.cleanupFn = cleanupFn;
    }
    InteractionWithPluginRequest.prototype.willRespondWith = function (status, builder) {
        this.interaction.withStatus(status);
        if (typeof builder === 'function') {
            builder(new responseWithPluginBuilder_1.ResponseWithPluginBuilder(this.interaction));
        }
        return new interactionWithPluginResponse_1.InteractionWithPluginResponse(this.pact, this.opts, this.cleanupFn);
    };
    return InteractionWithPluginRequest;
}());
exports.InteractionWithPluginRequest = InteractionWithPluginRequest;
