// @pact-foundation/pact/src/v4/http/interactionWithRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionWithRequest = void 0;
var interactionWithResponse_1 = require("./interactionWithResponse");
var responseBuilder_1 = require("./responseBuilder");
var InteractionWithRequest = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function InteractionWithRequest(pact, interaction, opts, cleanupFn) {
        this.pact = pact;
        this.interaction = interaction;
        this.opts = opts;
        this.cleanupFn = cleanupFn;
    }
    InteractionWithRequest.prototype.willRespondWith = function (status, builder) {
        this.interaction.withStatus(status);
        if (typeof builder === 'function') {
            builder(new responseBuilder_1.ResponseBuilder(this.interaction));
        }
        return new interactionWithResponse_1.InteractionWithResponse(this.pact, this.opts, this.cleanupFn);
    };
    return InteractionWithRequest;
}());
exports.InteractionWithRequest = InteractionWithRequest;
