// @pact-foundation/pact/src/v4/http/requestWithPluginBuilder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestWithPluginBuilder = void 0;
var requestBuilder_1 = require("./requestBuilder");
var RequestWithPluginBuilder = /** @class */ (function (_super) {
    __extends(RequestWithPluginBuilder, _super);
    function RequestWithPluginBuilder() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    RequestWithPluginBuilder.prototype.pluginContents = function (contentType, contents) {
        this.interaction.withPluginRequestInteractionContents(contentType, contents);
        return this;
    };
    RequestWithPluginBuilder.prototype.query = function (query) {
        _super.prototype.query.call(this, query);
        return this;
    };
    RequestWithPluginBuilder.prototype.headers = function (headers) {
        _super.prototype.headers.call(this, headers);
        return this;
    };
    RequestWithPluginBuilder.prototype.jsonBody = function (body) {
        _super.prototype.jsonBody.call(this, body);
        return this;
    };
    RequestWithPluginBuilder.prototype.binaryFile = function (contentType, file) {
        _super.prototype.binaryFile.call(this, contentType, file);
        return this;
    };
    /**
     * Sets the request body as multipart/form-data content for plugin-based interactions.
     * This is useful for testing APIs that accept file uploads or multipart form submissions.
     *
     * @param contentType - The content type of the multipart body (e.g., 'multipart/form-data')
     * @param file - Path to the file containing the multipart body content
     * @param mimePartName - The name of the mime part in the multipart body
     * @param boundary - Optional boundary string for the multipart content. If not provided, will be passed as undefined.
     * @returns The V4RequestWithPluginBuilder instance for method chaining
     */
    RequestWithPluginBuilder.prototype.multipartBody = function (contentType, file, mimePartName, boundary) {
        _super.prototype.multipartBody.call(this, contentType, file, mimePartName, boundary);
        return this;
    };
    /**
     * Applies matching rules to the consumer request.
     * Matching rules allow you to define flexible matching criteria for request attributes
     * beyond exact equality (e.g., regex patterns, type matching, number ranges).
     *
     * @param rules - The matching rules as a strongly typed Rules object. Rules should follow the Pact matching rules format.
     * @returns The V4RequestWithPluginBuilder instance for method chaining
     */
    RequestWithPluginBuilder.prototype.matchingRules = function (rules) {
        _super.prototype.matchingRules.call(this, rules);
        return this;
    };
    RequestWithPluginBuilder.prototype.body = function (contentType, body) {
        _super.prototype.body.call(this, contentType, body);
        return this;
    };
    return RequestWithPluginBuilder;
}(requestBuilder_1.RequestBuilder));
exports.RequestWithPluginBuilder = RequestWithPluginBuilder;
