// @pact-foundation/pact/src/v4/http/requestBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RequestBuilder = void 0;
var ramda_1 = require("ramda");
var matchers_1 = require("../../v3/matchers");
var _1 = require(".");
var matchingRules_1 = require("../../common/matchingRules");
var RequestBuilder = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function RequestBuilder(interaction) {
        this.interaction = interaction;
    }
    RequestBuilder.prototype.query = function (query) {
        var _this = this;
        (0, ramda_1.forEachObjIndexed)(function (v, k) {
            if (Array.isArray(v)) {
                v.forEach(function (vv, i) {
                    _this.interaction.withQuery(k, i, (0, matchers_1.matcherValueOrString)(vv));
                });
            }
            else {
                _this.interaction.withQuery(k, 0, (0, matchers_1.matcherValueOrString)(v));
            }
        }, query);
        return this;
    };
    RequestBuilder.prototype.headers = function (headers) {
        var _this = this;
        (0, ramda_1.forEachObjIndexed)(function (v, k) {
            if (Array.isArray(v)) {
                v.forEach(function (header, index) {
                    _this.interaction.withRequestHeader("".concat(k), index, (0, matchers_1.matcherValueOrString)(header));
                });
            }
            else {
                _this.interaction.withRequestHeader("".concat(k), 0, (0, matchers_1.matcherValueOrString)(v));
            }
        }, headers);
        return this;
    };
    RequestBuilder.prototype.jsonBody = function (body) {
        this.interaction.withRequestBody((0, matchers_1.matcherValueOrString)(body), 'application/json');
        return this;
    };
    RequestBuilder.prototype.binaryFile = function (contentType, file) {
        var body = (0, _1.readBinaryData)(file);
        this.interaction.withRequestBinaryBody(body, contentType);
        return this;
    };
    /**
     * Sets the request body as multipart/form-data content.
     * This is useful for testing APIs that accept file uploads or multipart form submissions.
     *
     * @param contentType - The content type of the multipart body (e.g., 'multipart/form-data')
     * @param file - Path to the file containing the multipart body content
     * @param mimePartName - The name of the mime part in the multipart body
     * @param boundary - Optional boundary string for the multipart content. If not provided, will be passed as undefined.
     * @returns The V4RequestBuilder instance for method chaining
     */
    RequestBuilder.prototype.multipartBody = function (contentType, file, mimePartName, boundary) {
        this.interaction.withRequestMultipartBody(contentType, file, mimePartName, boundary);
        return this;
    };
    /**
     * Applies matching rules to the consumer request.
     * Matching rules allow you to define flexible matching criteria for request attributes
     * beyond exact equality (e.g., regex patterns, type matching, number ranges).
     *
     * @param rules - The matching rules as a strongly typed Rules object. Rules should follow the Pact matching rules format.
     * @returns The V4RequestBuilder instance for method chaining
     */
    RequestBuilder.prototype.matchingRules = function (rules) {
        (0, matchingRules_1.validateRules)(rules);
        var ffiRules = (0, matchingRules_1.convertRulesToFFI)(rules);
        this.interaction.withRequestMatchingRules(JSON.stringify(ffiRules));
        return this;
    };
    RequestBuilder.prototype.body = function (contentType, body) {
        this.interaction.withRequestBinaryBody(body, contentType);
        return this;
    };
    return RequestBuilder;
}());
exports.RequestBuilder = RequestBuilder;
