// @pact-foundation/pact/src/v4/http/unconfiguredInteraction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnconfiguredInteraction = void 0;
var interactionWithPlugin_1 = require("./interactionWithPlugin");
var requestBuilder_1 = require("./requestBuilder");
var interactionWithRequest_1 = require("./interactionWithRequest");
var matchers_1 = require("../../v3/matchers");
var UnconfiguredInteraction = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function UnconfiguredInteraction(pact, interaction, opts, cleanupFn) {
        this.pact = pact;
        this.interaction = interaction;
        this.opts = opts;
        this.cleanupFn = cleanupFn;
    }
    UnconfiguredInteraction.prototype.uponReceiving = function (description) {
        this.interaction.uponReceiving(description);
        return this;
    };
    UnconfiguredInteraction.prototype.given = function (state, parameters) {
        if (parameters) {
            this.interaction.givenWithParams(state, JSON.stringify(parameters));
        }
        else {
            this.interaction.given(state);
        }
        return this;
    };
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    UnconfiguredInteraction.prototype.withCompleteRequest = function (request) {
        throw new Error('withCompleteRequest is not implemented');
    };
    UnconfiguredInteraction.prototype.withRequest = function (method, path, builder) {
        this.interaction.withRequest(method, (0, matchers_1.matcherValueOrString)(path));
        if (builder) {
            builder(new requestBuilder_1.RequestBuilder(this.interaction));
        }
        return new interactionWithRequest_1.InteractionWithRequest(this.pact, this.interaction, this.opts, this.cleanupFn);
    };
    UnconfiguredInteraction.prototype.usingPlugin = function (config) {
        this.pact.addPlugin(config.plugin, config.version);
        return new interactionWithPlugin_1.InteractionWithPlugin(this.pact, this.interaction, this.opts, this.cleanupFn);
    };
    return UnconfiguredInteraction;
}());
exports.UnconfiguredInteraction = UnconfiguredInteraction;
