// @pact-foundation/pact/src/v4/graphql/graphQLInteractionWithRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GraphQLInteractionWithRequest = void 0;
var lodash_1 = require("lodash");
var ramda_1 = require("ramda");
var graphQL_1 = require("../../common/graphQL/graphQL");
var interactionWithRequest_1 = require("../http/interactionWithRequest");
var types_1 = require("./types");
var matchers_1 = require("../../v3/matchers");
var GraphQLInteractionWithRequest = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function GraphQLInteractionWithRequest(pact, interaction, opts, cleanupFn, graphQLRequest) {
        this.pact = pact;
        this.interaction = interaction;
        this.opts = opts;
        this.cleanupFn = cleanupFn;
        this.graphQLRequest = graphQLRequest;
    }
    /**
     * The actual GraphQL query as a string.
     *
     * NOTE: spaces are not important, Pact will auto-generate a space-insensitive matcher
     *
     *  e.g. the value for the "query" field in the GraphQL HTTP payload:
     *  '{ "query": "{
     *        Category(id:7) {
     *          id,
     *          name,
     *          subcategories {
     *            id,
     *            name
     *          }
     *        }
     *     }"
     *  }'
     * @param query {string|ASTNode} parsed or unparsed query
     * @return this object
     */
    GraphQLInteractionWithRequest.prototype.withQuery = function (query) {
        return this.setQueryDetails(query, types_1.OperationType.Query);
    };
    GraphQLInteractionWithRequest.prototype.withMutation = function (mutation) {
        return this.setQueryDetails(mutation, types_1.OperationType.Mutation);
    };
    GraphQLInteractionWithRequest.prototype.setQueryDetails = function (query, type) {
        var validatedQuery = (0, graphQL_1.validateQuery)(query, type);
        this.interaction.withRequestBody(JSON.stringify((0, ramda_1.reject)(lodash_1.isUndefined, {
            operationName: this.graphQLRequest.operation,
            query: (0, matchers_1.regex)((0, graphQL_1.escapeGraphQlQuery)(validatedQuery), validatedQuery),
            variables: this.graphQLRequest.variables,
        })), 'application/json');
        return new interactionWithRequest_1.InteractionWithRequest(this.pact, this.interaction, this.opts, this.cleanupFn);
    };
    return GraphQLInteractionWithRequest;
}());
exports.GraphQLInteractionWithRequest = GraphQLInteractionWithRequest;
