// @pact-foundation/pact/src/v4/graphql/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnconfiguredGraphQLInteraction = void 0;
var matchers_1 = require("../../v3/matchers");
var graphQLRequestBuilder_1 = require("./graphQLRequestBuilder");
var graphQLInteractionWithRequest_1 = require("./graphQLInteractionWithRequest");
var UnconfiguredGraphQLInteraction = /** @class */ (function () {
    // tslint:disable:no-empty-function
    function UnconfiguredGraphQLInteraction(pact, interaction, opts, cleanupFn) {
        this.pact = pact;
        this.interaction = interaction;
        this.opts = opts;
        this.cleanupFn = cleanupFn;
        this.graphQLRequest = {};
    }
    UnconfiguredGraphQLInteraction.prototype.withOperation = function (operation) {
        this.graphQLRequest.operation = operation;
        return this;
    };
    UnconfiguredGraphQLInteraction.prototype.withVariables = function (variables) {
        this.graphQLRequest.variables = variables;
        return this;
    };
    UnconfiguredGraphQLInteraction.prototype.uponReceiving = function (description) {
        this.interaction.uponReceiving(description);
        return this;
    };
    UnconfiguredGraphQLInteraction.prototype.given = function (state, parameters) {
        if (parameters) {
            this.interaction.givenWithParams(state, JSON.stringify(parameters));
        }
        else {
            this.interaction.given(state);
        }
        return this;
    };
    UnconfiguredGraphQLInteraction.prototype.withRequest = function (method, path, builder) {
        this.interaction.withRequest(method, (0, matchers_1.matcherValueOrString)(path));
        if (builder) {
            builder(new graphQLRequestBuilder_1.GraphQLRequestBuilder(this.interaction));
        }
        return new graphQLInteractionWithRequest_1.GraphQLInteractionWithRequest(this.pact, this.interaction, this.opts, this.cleanupFn, this.graphQLRequest);
    };
    return UnconfiguredGraphQLInteraction;
}());
exports.UnconfiguredGraphQLInteraction = UnconfiguredGraphQLInteraction;
