// @pact-foundation/pact/src/dsl/verifier/proxy/messages.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupMessageProxyApplication = exports.createProxyMessageHandler = exports.providerWithMetadata = exports.findMessageHandler = void 0;
var express_1 = __importDefault(require("express"));
var body_parser_1 = __importDefault(require("body-parser"));
var js_base64_1 = require("js-base64");
var logger_1 = __importDefault(require("../../../common/logger"));
// Find a provider message handler, and invoke it
var findMessageHandler = function (message, config) {
    var handler = config.messageProviders
        ? config.messageProviders[message.description]
        : undefined;
    if (!handler) {
        logger_1.default.error("no handler found for message ".concat(message.description));
        return Promise.reject(new Error("No handler found for message \"".concat(message.description, "\".\n             Check your \"handlers\" configuration")));
    }
    return Promise.resolve(handler);
};
exports.findMessageHandler = findMessageHandler;
var hasMetadata = function (o) {
    return Boolean(o.__pactMessageMetadata);
};
var providerWithMetadata = function (provider, metadata) {
    return function (descriptor) {
        return Promise.resolve(provider(descriptor)).then(function (message) {
            return hasMetadata(message)
                ? {
                    __pactMessageMetadata: __assign(__assign({}, message.__pactMessageMetadata), metadata),
                    message: message,
                }
                : { __pactMessageMetadata: metadata, message: message };
        });
    };
};
exports.providerWithMetadata = providerWithMetadata;
// Get the API handler for the verification CLI process to invoke on POST /*
var createProxyMessageHandler = function (config) {
    return function (req, res) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        // req.body = {
        //   description: 'a MATT message',
        //   providerStates: [ { name: 'a Matt message is sent' } ],
        //   request: {
        //     contents: {
        //       content: 'aGVsbG90Y3A=',
        //       contentType: 'application/matt',
        //       encoded: 'base64'
        //     },
        //     matchingRules: {
        //       body: {
        //         '$': {
        //           combine: 'AND',
        //           matchers: [ { match: 'contentType', value: 'application/matt' } ]
        //         }
        //       }
        //     },
        //     metadata: { contentType: 'application/matt' }
        //   }
        // }
        // {
        //   description: 'a MATT message',
        //   providerStates: [ { name: 'a Matt message is sent with JSON content' } ],
        //   request: {
        //     contents: {
        //       content: { message: 'hellotcp' },
        //       contentType: 'application/json',
        //       encoded: false
        //     },
        //     metadata: { contentType: 'application/json' }
        //   }
        // }
        var message = req.body;
        message.metadata = (_a = req.body.request) === null || _a === void 0 ? void 0 : _a.metadata;
        // Add the request body, if one exists for synchronous messages
        if ((_b = req.body.request) === null || _b === void 0 ? void 0 : _b.contents) {
            var encodedMessage = ((_d = (_c = req.body.request) === null || _c === void 0 ? void 0 : _c.contents) === null || _d === void 0 ? void 0 : _d.encoded) || false;
            if (encodedMessage) {
                message.content = Buffer.from((_f = (_e = req.body.request) === null || _e === void 0 ? void 0 : _e.contents) === null || _f === void 0 ? void 0 : _f.content, 'base64');
            }
            else {
                message.content = (_h = (_g = req.body.request) === null || _g === void 0 ? void 0 : _g.contents) === null || _h === void 0 ? void 0 : _h.content;
            }
        }
        // Invoke the handler, and return the JSON response body
        // wrapped in a Message
        (0, exports.findMessageHandler)(message, config)
            .then(function (handler) { return handler(message); })
            .then(function (messageFromHandler) {
            if (hasMetadata(messageFromHandler)) {
                var metadata = (0, js_base64_1.encode)(JSON.stringify(messageFromHandler.__pactMessageMetadata));
                res.header('Pact-Message-Metadata', metadata);
                res.header('PACT_MESSAGE_METADATA', metadata);
                // Ensure return content-type is set here
                // TODO: this needs thinking through
                if (messageFromHandler.__pactMessageMetadata['content-type'] ||
                    messageFromHandler.__pactMessageMetadata['Content-Type'] ||
                    messageFromHandler.__pactMessageMetadata.contentType) {
                    res.type(messageFromHandler.__pactMessageMetadata['content-type'] ||
                        messageFromHandler.__pactMessageMetadata['Content-Type'] ||
                        messageFromHandler.__pactMessageMetadata.contentType);
                    return res.send(messageFromHandler.message);
                }
                return res.json(messageFromHandler.message);
            }
            return res.json(messageFromHandler);
        })
            .catch(function (e) { return res.status(500).send(e); });
    };
};
exports.createProxyMessageHandler = createProxyMessageHandler;
// Get the Express app that will run on the HTTP Proxy
var setupMessageProxyApplication = function (config) {
    var app = (0, express_1.default)();
    app.use(body_parser_1.default.json());
    app.use(body_parser_1.default.urlencoded({ extended: true }));
    app.use(function (_, res, next) {
        // TODO: this seems to override the metadata for content-type
        res.header('Content-Type', 'application/json; charset=utf-8');
        next();
    });
    // Proxy server will respond to Verifier process
    app.all('/{*splat}', (0, exports.createProxyMessageHandler)(config));
    return app;
};
exports.setupMessageProxyApplication = setupMessageProxyApplication;
// // Get the Proxy we'll pass to the CLI for verification
// export const setupProxyServer = (
//   app: (request: http.IncomingMessage, response: http.ServerResponse) => void
// ): http.Server => http.createServer(app).listen();
