// @pact-foundation/pact/src/v4/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Pact = exports.PactV4 = void 0;
var pact_core_1 = require("@pact-foundation/pact-core");
var unconfiguredInteraction_1 = require("./http/unconfiguredInteraction");
var package_json_1 = require("../../package.json");
var message_1 = require("./message");
var v3_1 = require("../v3");
var graphql_1 = require("./graphql");
var asynchronousMessage_1 = require("./message/asynchronousMessage");
var PactV4 = /** @class */ (function () {
    function PactV4(opts) {
        this.opts = opts;
        this.setup();
        this.pact.addMetadata('pact-js', 'version', package_json_1.version);
    }
    PactV4.prototype.setup = function () {
        var _a, _b;
        this.pact = (0, pact_core_1.makeConsumerPact)(this.opts.consumer, this.opts.provider, (_a = this.opts.spec) !== null && _a !== void 0 ? _a : v3_1.SpecificationVersion.SPECIFICATION_VERSION_V4, (_b = this.opts.logLevel) !== null && _b !== void 0 ? _b : 'info');
    };
    PactV4.prototype.addInteraction = function () {
        var _this = this;
        return new unconfiguredInteraction_1.UnconfiguredInteraction(this.pact, this.pact.newInteraction(''), this.opts, function () {
            // This function needs to be called if the PactV4 object is to be re-used (commonly expected by users)
            // Because of the type-state model used here, it's a bit awkward as we need to thread this through
            // to children, ultimately to be called on the "executeTest" stage.
            _this.setup();
        });
    };
    PactV4.prototype.addSynchronousInteraction = function (description) {
        var _this = this;
        return new message_1.UnconfiguredSynchronousMessage(this.pact, this.pact.newSynchronousMessage(description), this.opts, function () {
            // This function needs to be called if the PactV4 object is to be re-used (commonly expected by users)
            // Because of the type-state model used here, it's a bit awkward as we need to thread this through
            // to children, ultimately to be called on the "executeTest" stage.
            _this.setup();
        });
    };
    PactV4.prototype.addAsynchronousInteraction = function () {
        var _this = this;
        return new asynchronousMessage_1.UnconfiguredAsynchronousMessage(this.pact, this.pact.newAsynchronousMessage(''), this.opts, function () {
            // This function needs to be called if the PactV4 object is to be re-used (commonly expected by users)
            // Because of the type-state model used here, it's a bit awkward as we need to thread this through
            // to children, ultimately to be called on the "executeTest" stage.
            _this.setup();
        });
    };
    PactV4.prototype.addGraphQLInteraction = function () {
        var _this = this;
        return new graphql_1.UnconfiguredGraphQLInteraction(this.pact, this.pact.newInteraction(''), this.opts, function () {
            // This function needs to be called if the PactV4 object is to be re-used (commonly expected by users)
            // Because of the type-state model used here, it's a bit awkward as we need to thread this through
            // to children, ultimately to be called on the "executeTest" stage.
            _this.setup();
        });
    };
    return PactV4;
}());
exports.PactV4 = PactV4;
exports.Pact = PactV4;
__exportStar(require("./graphql"), exports);
__exportStar(require("./message/index"), exports);
